code/obj/machinery/turret.dm

```diff
@@ -379,74 +379,55 @@
 /obj/machinery/turretid/attack_ai(mob/user as mob)
 	return attack_hand(user)
 
-/obj/machinery/turretid/attack_hand(mob/user as mob)
-	if (user.getStatusDuration("stunned") || user.getStatusDuration("weakened") || user.stat)
-		return
-
-	if(!in_interact_range(src, user))
-		boutput(user, text("Too far away."))
-		src.remove_dialog(user)
-		user.Browse(null, "window=turretid")
-		return
-
-	src.add_dialog(user)
+/obj/machinery/turretid/ui_interact(mob/user, datum/tgui/ui)
+	if (src.locked)
+		if (!issilicon(usr) && !isAI(usr))
+			boutput(usr, "Control panel is locked!")
+			return
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "TurretControl")
+		ui.open()
+
+/obj/machinery/turretid/ui_data(mob/user)
+	. = list(
+		"enabled" = src.enabled,
+		"lethal" = src.lethal,
+		"emagged" = src.emagged
+	)
+
+/obj/machinery/turretid/ui_static_data(mob/user)
 	var/area/area = get_area(src)
 	if (!istype(area))
 		logTheThing("debug", null, null, "Turret badly positioned.")
-		return
-	var/t = "<TT><B>Turret Control Panel</B> ([area.name])<HR>"
-
-	if(!src.emagged)
-		if(src.locked && (!issilicon(user) && !isAI(user)))
-			t += "<I>(Swipe ID card to unlock control panel.)</I><BR>"
```

Comment:
Looks like the old panel had some functionality to show if it was locked or not, could we keep that in the UI rather than having it be when someone tries to open the panel? A good example of an issue here would be if I (as a human) open an unlocked panel then someone locks it, there is no visual indication to me that it was locked.